File: real-life/scalaCodingChallenges/src/main/scala/scalaCodingChallenges/MaxSubListDistinctValues.scala
package scalaCodingChallenges

object MaxSubListDistinctValues extends App {
  // TODO: Return the Max() number of Distinct() values, among the rolling consecutive subList-s of size M,
  //  example_01: List(1, 1, 3, 3, 5, 3), N=6, M=3 -> 2
  //  example_02: List("A", "A", "C", "C", "E", "C"), N=6, M=3 -> 2

  val values = List("A", "A", "C", "C", "E", "C")
  val n = values.size
  val m = 3

  println (maxSubListDistinctValues(values, n, m))

  private def maxSubListDistinctValues[T](values: List[T], n: Int, m: Int) = ???

}


//return IntStream.rangeClosed(0, n - m)
//.parallel()
//.mapToObj(i -> values.subList(i, i + m)
//.stream()
//.distinct()
//.count()
//)
//.reduce(Long::max)
//.orElse(0L);

File: real-life/javaCodingChallenges/src/main/java/codingChallenges/MaxSubListDistinctValues.java
package codingChallenges;

import java.util.Arrays;
import java.util.List;
import java.util.stream.IntStream;

public class MaxSubListDistinctValues {
    // TODO: Return the Max() number of Distinct() values,
    //  among the rolling consecutive subList-s of size M, 
    //  for a given List of N Integers
    //  example: List(1, 1, 3, 3, 5, 3), N=6, M=3 -> 2

    public static void main(String[] args) {
        List<Integer> values = Arrays.asList(1, 1, 3, 3, 5, 3);
        int N = values.size();
        int M = 3;

        System.out.println(maxSubListDistinctValues(values, N, M));
    }

    public static long maxSubListDistinctValues(List<Integer> values, int N, int M) {
        return IntStream.rangeClosed(0, N - M).parallel()
                .mapToObj(i -> values.subList(i, i + M).stream()
                        .distinct()
                        .count()
                )
                .reduce(Long::max)
                .get();
    }
}